### PR TITLE
[TRTLLM-3442] feat: added beam search support to the PyTorch Workflow

### PR DESCRIPTION
**Pull request by stnie**
## Description

- Added Basic Beam Search support for the PyTorch Workflow. 
- Return Context Logits and Generation Logits support for beam search is provided
- Return Logprobs Support is only provided for a single prompt
- Tests were run using TinyLlama model 

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by dcampora — CHANGES_REQUESTED** [review]
Looks good to me, please do the changes required and let's merge it.

File: tensorrt_llm/_torch/pyexecutor/resource_manager.py [hunk 1]
```diff
@@ -333,7 +334,7 @@ def add_dummy_requests(
         prepare_resource: bool = True,
         max_num_draft_tokens: int = 0,
     ):
-        beam_width = 1
+        beam_width = 1  # TODO: more than 1 beam?
```
[the remark this answers is not in the document]
Comment (stnie) [reply tensorrt_llm/_torch/pyexecutor/resource_manager.py:337]: During warmup, I think it can stay at 1.

File: tensorrt_llm/_torch/pyexecutor/sampler.py [hunk 2]
```diff
@@ -762,3 +783,81 @@ def update_requests(self, state: SampleStateTRTLLM):
...
+        output_ids_host = self.store["decoder_state"].gathered_ids.to(
+            'cpu', non_blocking=False)
+        sequence_lengths_host = self.store["decoder_state"].sequence_lengths.to(
+            'cpu', non_blocking=False)
+
+        if request.py_return_log_probs:
+            log_probs_host = self.store["decoder_state"].log_probs.to(
+                'cpu', non_blocking=False)
+            cum_log_probs_host = self.store["decoder_state"].cum_log_probs.to(
+                'cpu', non_blocking=False)
```
[the remark this answers is not in the document]
Comment (stnie) [reply tensorrt_llm/_torch/pyexecutor/sampler.py]: done (here and in other cases)

File: tests/unittest/_torch/test_beam_search.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+@pytest.mark.parametrize("num_prompts", [1, 2])
+def test_beam_search_output_shapes(disable_overlap_scheduler: bool,
+                                   enable_trtllm_sampler: bool,
+                                   gather_context_logits: bool,
+                                   gather_generation_logits: bool,
+                                   return_log_probs: bool, max_beam_width: int,
+                                   max_tokens: int, num_prompts: int):
+    if not enable_trtllm_sampler:
+        pytest.skip(
+            "Beam search currently is only supported with TRTLLMSampler")
```
[the remark this answers is not in the document]
Comment (stnie) [reply tests/unittest/_torch/test_beam_search.py]: done

**Review by Funatiq — CHANGES_REQUESTED** [review]
Do we have a check somewhere for `executor_config.max_beam_width == req_beam_width`? What happens if this is not the case? It would be good to have a test for that.

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 4]
```diff
@@ -482,6 +482,12 @@ def is_nvfp4_output_kernel_available(
 class TrtllmAttentionMetadata(AttentionMetadata):
     workspace: Optional[torch.Tensor] = None
 
+    # TrtllmAttention needs to know the beam width and access to the cache indirection buffer,
+    # when beam search is enabled.
+    # TODO -- put this into the parent class instead?
```
[the remark this answers is not in the document]
Comment (stnie) [reply tensorrt_llm/_torch/attention_backend/trtllm.py]: I left in in the TrtllmAttentionMetadata and removed the todo

File: tensorrt_llm/_torch/pyexecutor/llm_request.py [hunk 5]
```diff
@@ -138,6 +138,16 @@ def append(self,
...
                     next(iter(prob.values())).logprob for prob in probs)
 
+    def set_log_probs(self, log_probs: list[TokenLogprobs],
+                      cum_log_probs: list[float]):
+        """
+        Reset the storage and refill it with new values
+        log_probs: [beam_width, num_tokens]
+        cum_log_probs: [beam_width]
+        """
+        self.beam_width = -1
```
[the remark this answers is not in the document]
Comment (stnie) [reply tensorrt_llm/_torch/pyexecutor/llm_request.py]: I previously relied on the fact, that append, reinitializes the buffer if beam_width is -1. However, for clarity I directly call the init function instead of setting beam_width to -1.

**Review by Funatiq — APPROVED** [review]

File: tests/integration/test_lists/test-db/l0_a30.yml [hunk 6]
```diff
@@ -18,6 +18,7 @@ l0_a30:
   - unittest/_torch/modeling -k "modeling_qwen"
   - unittest/_torch/modeling -k "modeling_qwen_moe"
   - unittest/_torch/auto_deploy/unit/singlegpu
+  - unittest/_torch/test_beam_search.py
```
Comment (dcampora) [inline tests/integration/test_lists/test-db/l0_a30.yml:21]: No need to add it here.

Comment (Funatiq) [reply tests/integration/test_lists/test-db/l0_a30.yml]: IIRC @stnie mentioned that it doesn't run if it isn't added. Have we confirmed that it runs in another stage?

**Review by dcampora — APPROVED** [review]

**Review by suyoggupta — APPROVED** [review]
